### PR TITLE
Fill out TrackSegment API

### PR DESCRIPTION
**Pull request by spacek531**
Exposes track properties of TED to scripting API.

I would like feedback on the implementation of TrackCurveType and TrackSlopeType.

I thought about exposing gTrackDescriptors but I changed my mind - plugin makers can decide for themselves how best to look up track elements to build.

I could add sequence specific data (allowed walls, entrance hut directions, if a height marker is shown), but I don't see a use for it at this time.

### CODE REVIEW

File: distribution/openrct2.d.ts
Comment (spacek531) [inline distribution/openrct2.d.ts:1238]: The only user of this value is when calculating the circuit length to determine the number of powered vehicles allowed on some ride types (like log flume). It is in no way accurate to the length of the piece - that data is the subposition data.

File: distribution/openrct2.d.ts
Comment (spacek531) [inline distribution/openrct2.d.ts:1217]: Negative numbers indicate the starting position is below the height of the first track sequence block. CableLiftHill, FlatToDown60LongBase, Down60ToFlatLongBase have negative endZ values.

File: src/openrct2/scripting/bindings/game/ScContext.hpp
Comment (spacek531) [inline src/openrct2/scripting/bindings/game/ScContext.hpp:235]: I did this instead of exposing a property to context with the number of track elements. I could still add that too.

**Comment by spacek531** [comment]
@Basssiiie @Sadret @oli414 @IntelOrca Do you have any comments or suggestions?

File: distribution/openrct2.d.ts
Comment (ltsSmitty) [inline distribution/openrct2.d.ts:1301]: is there another accessor for isSteepDown? An alternate exposure could be "isSteep" + TrackCurveType.Down

Comment (ltsSmitty) [reply distribution/openrct2.d.ts:1301]: Same line of thinking, how do vertical track pieces get exposed in this schema?

Comment (spacek531) [reply distribution/openrct2.d.ts:1301]: There are no accompanying values for isSteepUp, nor is there anything for any other slope angle. This flag is present for the sole purpose of limiting steep chain lifts to reverse sections for reverse incline shuttle mode but disallowing steep chain lifts on forward pieces.

File: distribution/openrct2.d.ts
Comment (spacek531) [reply distribution/openrct2.d.ts]: They refer to the next track piece that is selected when you build a track piece... for example large half loop left down is automatically selected when you build a large half loop up. I implemented Basssiiie's suggestion regarding name and values.

File: distribution/openrct2.d.ts
Comment (ltsSmitty) [reply distribution/openrct2.d.ts:1296]: that's a great comment to add. thanks.

File: distribution/openrct2.d.ts
Comment (ltsSmitty) [inline distribution/openrct2.d.ts]: I don't think `chainEndCurve` means anything anymore. Can you update this comment?

Comment (spacek531) [reply distribution/openrct2.d.ts]: Weird, it's fixed on my end, I thought I pushed

**Comment by Basssiiie** [comment]
Regarding the `countsAsInversion` and similar flags; is there for example an inversion that does not count as an inversion? If all inversions have the flag, then I'd say `isInversion` is fine. If some have the flag and some others don't, then I'd say `countsAsInversion` is better. The `countsAs*` seems to imply some don't count as that.

**Comment by spacek531** [comment]
The first half of half loops, corkscrews, barrel rolls, and quarter loops (i.e. LeftBarrelRollUpToDown) have `isInversion` and `countsAsInversion` but the second half (i.e. RightBarrelRollDownToUp) do not have `countsAsInversion`.

**Comment by Basssiiie** [comment]
Ah alright, then for `countsAsGolfHole` and `countsAsHelix`, are there any golf holes or helixes without this flag?

**Comment by spacek531** [comment]
as far as I know, no. countsAsGolfHole is actually an alias of countsAsInversion since they occupy the same bit (and work the same way in the game code). I chose this naming schema because they have identical underlying behavior - they all add to the rating bins.

**Comment by Basssiiie** [comment]
Ah lol so a looping would also count as a golf hole? Then I'd suggest to keep `countsAsGolfHole` as is but `countsAsHelix` could be changed to `isHelix`.  

Because I don't think how it is used internally is very important for the naming of the property. It's more important that what it could return for any track segments is somewhat intuitive and expected.

**Comment by spacek531** [comment]
I've made some changes. All the "countsAs" flags are now "is" and the `isBanked` and `isInversion` flags are now `banked` and `inverts`.

File: src/openrct2/scripting/ScriptEngine.h
Comment (spacek531) [inline src/openrct2/scripting/ScriptEngine.h:49]: Is this necessary if API is only added to?

Comment (IntelOrca) [reply src/openrct2/scripting/ScriptEngine.h:49]: It allows plugins to easily check if one or more changes are available / present.

Comment (spacek531) [reply src/openrct2/scripting/ScriptEngine.h:49]: So, yes?

Comment (Basssiiie) [reply src/openrct2/scripting/ScriptEngine.h:49]: Yes

**Comment by Basssiiie** [comment]
> I've made some changes. All the "countsAs" flags are now "is" and the `isBanked` and `isInversion` flags are now `segmentBanks` and `segmentInverts` with description of the exact behaviour they represent in the documentation.

Tbh I preferred the `CountAs` flags for the flags that do not represent the whole exact set of track pieces. Now `isInversion` will be false for half the inversion pieces and `isGolfHole` will be true for loopings and corkscrews. That can be quite confusing. The `CountAs` flags activily imply some inversions may not be counted as inversions.

**Comment by spacek531** [comment]
Fair enough, I am convinced.

**Review by Basssiiie — APPROVED** [review]
LGTM, great work 😃

File: src/openrct2/scripting/bindings/ride/ScTrackSegment.cpp
Comment (duncanspumpkin) [inline src/openrct2/scripting/bindings/ride/ScTrackSegment.cpp:35]: I'd much rather they were exposed as yaw, roll, pitch rather than direction, bank and angle. yaw, roll, pitch are much clearer as to what they mean compared to direction, bank and angle.

Comment (spacek531) [reply src/openrct2/scripting/bindings/ride/ScTrackSegment.cpp:35]: They were named this by whoever added them... I haven't done git-blame to see who. I will change them.

Comment (Basssiiie) [reply src/openrct2/scripting/bindings/ride/ScTrackSegment.cpp:35]: Please don't change the names already exposed to the plugin developers. It would be a breaking change that would break existing plugins.

**Comment by spacek531** [comment]
I've changed Angle to Pitch. I didn't change Direction because it was already present in the API.

**Comment by ltsSmitty** [comment]
are there any remaining blockers on this? I'm at a point in my plugin development that I can't really move on til this is implemented, so curious about next steps

**Comment by Sadret** [comment]
> are there any remaining blockers on this? I'm at a point in my plugin development that I can't really move on til this is implemented, so curious about next steps

You can download the build for your local plugin development, without waiting for it to be merged.

**Comment by spacek531** [comment]
I've added a reverse map when getting the widget's icon, for completeness.

**Comment by tupaschoal** [comment]
As far as I see, this just needs to have merge conflicts solved and a re-review by Duncan.

**Comment by spacek531** [comment]
I have rebased this. I do not know why github flagged it as merge conflicts because I did not have to fix anything.

**Comment by tupaschoal** [comment]
@duncanspumpkin can you rereview this one, please?

**Comment by duncanspumpkin** [comment]
What happened with the angle, pitch rename? I see you've gone for slope. I can't see why that was chosen the last message in the chain said it was going to change to pitch.

**Comment by spacek531** [comment]
Pitch and slope seemed equivalent to me

**Comment by spacek531** [comment]
I remember why I named it slope instead of pitch. The existing enum TrackSlope is called slope, and start/endPitch uses that enum.

**Comment by duncanspumpkin** [comment]
> I remember why I named it slope instead of pitch. The existing enum TrackSlope is called slope, and start/endPitch uses that enum.

argh didn't realise the name slope was already in use. Hmm. Maybe be best to keep it slope then.

**Comment by spacek531** [comment]
ok, I will revert the commit

**Comment by duncanspumpkin** [comment]
Sorry for all the little comments but could you bump the plugin version so that users can have a min target for when this exists.

**Review by duncanspumpkin — APPROVED** [review]

**Comment by spacek531** [comment]
plugin API is bumped

**Pull request merged**
